File: 02/password_validator.rb
def readFile(filename)
    file = File.open(filename)
    rulesWithPasswords = file.readlines.map(&:chomp)
end

def parsePassword(line)
    rule, password = line.split(":")
    bounds, character = rule.split(" ")
    lowerBound, upperBound = bounds.split("-").map(&:to_i)
    return password.strip(), character, lowerBound, upperBound
end

def part1
    rulesWithPasswords = readFile("input")
    validPasswords = []
    rulesWithPasswords.each do |ruleWithPassword|
        password, character, lowerBound, upperBound = parsePassword(ruleWithPassword)
        if lowerBound <= password.count(character) && password.count(character) <= upperBound
            validPasswords.push(ruleWithPassword)
        end
    end

    puts "Part 1"
    puts "Valid Passwords#{validPasswords}"
    puts "Total:#{validPasswords.count}"
end

def part2
    rulesWithPasswords = readFile("input")
    validPasswords = []
    rulesWithPasswords.each do |ruleWithPassword|
        password, character, lowerPosition, upperPosition = parsePassword(ruleWithPassword)
        if (password[lowerPosition-1] == character && password[upperPosition-1] != character) || 
            (password[lowerPosition-1] != character && password[upperPosition-1] == character)
            validPasswords.push(ruleWithPassword)
        end
    end

    puts "Part 2"
    puts "Valid Passwords#{validPasswords}"
    puts "Total:#{validPasswords.count}"
end

part2

File: 02/password_validator.ex
defmodule PasswordValidator do
    def readRulesFromFile() do
        {:ok, body} = File.read("input")
        body 
            |> String.split("\n") 
    end

    def parseRuleWithPassword(ruleWithPassword) do
        [rule, password] = String.split(ruleWithPassword, ":")
        [bounds, character] = String.split(rule, " ")
        [lower, upper] = Enum.map(String.split(bounds, "-"), &String.to_integer/1)
        %{ password: String.trim(password), character: character, lower: lower, upper: upper }
    end

    def characterCount(string, character) do 
        string
            |> String.graphemes
            |> Enum.count(&(&1 == character))
    end

    def isValidPart1?(%{character: character, password: password, lower: lower, upper: upper}) do
        characterCount = characterCount(password, character)
        lower <= characterCount && characterCount <= upper
    end

    def isValidPart2?(%{character: character, password: password, lower: lower, upper: upper}) do
        (String.at(password, lower-1) == character && String.at(password, upper-1) != character) || 
            (String.at(password, lower-1) != character && String.at(password, upper-1) == character)
    end

    def validate([]) do
        []
    end

    def validate(rulesWithPasswords) do
        [ head | tail ] = rulesWithPasswords
        Enum.filter([ parseRuleWithPassword(head) | validate(tail) ], &isValidPart2?/1)
    end

    def countValidPasswords do
        readRulesFromFile()
            |> validate
            |> Enum.count
    end
end

IO.inspect PasswordValidator.countValidPasswords, label: "Valid Passwords"